### PR TITLE
Multidimensional reward

### LINKED ISSUE #615: Vector reward
In many RL problems, the scalar reward is the sum of several components. It may be useful to predict the reward from each components separately. There are two potential benefits of doing this:

1. More signal for training the internal representation.
2. It allows the possibility of adjusting the contribution of each component to the final scalar reward.

### PR DESCRIPTION
**Pull request by emailweixu**
Also merge OneStepTDLoss with MultiStepTDLoss

### CODE REVIEW

File: alf/algorithms/ddpg_algorithm.py [hunk 1]
```diff
@@ -245,6 +245,11 @@ def _critic_train_step(self, exp: Experience, state: DdpgCriticState):
         target_q_values, target_critic_states = self._target_critic_networks(
             (exp.observation, target_action), state=state.target_critics)
 
+        if self._num_critic_replicas > 1:
+            target_q_values = target_q_values.min(dim=1)[0]
+        else:
+            target_q_values = target_q_values.squeeze(dim=1)
```
Comment (hnyu) [inline alf/algorithms/ddpg_algorithm.py:260]: These two can be unified as min(dim=1)[0]?

File: alf/algorithms/ddpg_algorithm.py [hunk 2]
```diff
@@ -264,7 +269,13 @@ def _actor_train_step(self, exp: Experience, state: DdpgActorState):
 
         q_values, critic_states = self._critic_networks(
             (exp.observation, action), state=state.critics)
-        q_value = q_values.min(dim=1)[0]
+        if q_values.ndim == 3:
+            # Multidimensional reward: [B, num_criric_replicas, reward_dim]
+            q_values = q_values.sum(dim=2)
```
Comment (hnyu) [inline alf/algorithms/ddpg_algorithm.py]: I think we need to reserve a place for multiplying values with the corresponding reward weights.

File: alf/algorithms/one_step_loss.py [hunk 3]
```diff
@@ -21,15 +21,16 @@
...
 
 @gin.configurable
-class OneStepTDLoss(nn.Module):
+class OneStepTDLoss(TDLoss):
     def __init__(self,
                  gamma=0.99,
                  td_error_loss_fn=losses.element_wise_squared_loss,
                  debug_summaries=False,
-                 name="OneStepLoss"):
+                 name="TDLoss"):
```
Comment (hnyu) [inline alf/algorithms/one_step_loss.py:33]: The name should still be "OneStepTDLoss"?

File: alf/algorithms/td_loss.py [hunk 4]
```diff
@@ -106,15 +113,30 @@ def forward(self, experience, value, target_value):
...
+                    for i in range(value.shape[2]):
+                        suffix = '/' + str(i)
+                        _summarize(value[..., i], returns[..., i], td[..., i],
+                                   suffix)
 
         loss = self._td_error_loss_fn(returns.detach(), value)
 
+        if loss.ndim == 3:
+            # Multidimensional reward. Average over the critic loss for all dimensions
+            loss = loss.mean(dim=2)
```
Comment (hnyu) [inline alf/algorithms/td_loss.py:138]: Should we use mean or sum? If training critics of each reward is independent with each other, then sum is probably better.

File: alf/data_structures.py [hunk 5]
```diff
@@ -397,7 +397,7 @@ def termination(observation, prev_action, reward, env_id=None, env_info={}):
             env_info=env_info)
 
 
-def time_step_spec(observation_spec, action_spec):
+def time_step_spec(observation_spec, action_spec, reward_spec):
     """Returns a ``TimeStep`` spec given the ``observation_spec`` and the
     ``action_spec``.
```
Comment (hnyu) [inline alf/data_structures.py:402]: Comment needs to be updated to include reward_spec

File: alf/environments/suite_unittest.py [hunk 6]
```diff
@@ -170,6 +182,10 @@ def _gen_time_step(self, s, action):
                                      action.reshape(prev_observation.shape))
             reward = reward.reshape(self.batch_size)
 
+        if self._reward_dim != 1:
+            reward = reward.unsqueeze(-1).expand((self.batch_size,
```
Comment (hnyu) [inline alf/environments/suite_unittest.py]: self.batch_size can just be -1

File: alf/utils/value_ops.py [hunk 7]
```diff
@@ -180,6 +184,11 @@ def one_step_discounted_return(rewards, values, step_types, discounts):
                                       s=values.shape[0]))
 
     is_lasts = (step_types == StepType.LAST).to(dtype=torch.float32)
+    if is_lasts.ndim < values.ndim:
```
Comment (hnyu) [inline alf/utils/value_ops.py]: suggest putting this ndim check inside expand_dim_as to simplify code

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/networks/critic_networks.py [hunk 8]
```diff
@@ -80,6 +81,7 @@ def __init__(self,
         Args:
             input_tensor_spec: A tuple of ``TensorSpec``s ``(observation_spec, action_spec)``
                 representing the inputs.
+            ourput_tensor_spec (TensorSpec): spec for the output
```
Comment (Haichao-Zhang) [inline alf/networks/critic_networks.py]: output_tensor_spec

File: alf/algorithms/ddpg_algorithm.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/ddpg_algorithm.py:260]: min is slow, squeeze is faster since it does not need to do any real calculation.

File: alf/algorithms/ddpg_algorithm.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/ddpg_algorithm.py]: done

File: alf/algorithms/one_step_loss.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/one_step_loss.py:33]: I intentionally changed so name so that the curves can be shown together if the loss is switched.

File: alf/algorithms/td_loss.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/td_loss.py:138]: The current implementation of critic network trains all rewards together. Even if we have separate critic network for each reward, mean() or sum() doesn't matter much because Adam optimizer is largely invariant of it.

File: alf/data_structures.py (hunk 5 above)
Comment (emailweixu) [reply alf/data_structures.py:402]: updated.

File: alf/utils/value_ops.py (hunk 7 above)
Comment (emailweixu) [reply alf/utils/value_ops.py]: removed.

File: alf/environments/suite_unittest.py (hunk 6 above)
Comment (emailweixu) [reply alf/environments/suite_unittest.py]: changed.

File: alf/utils/value_ops.py [hunk 9]
```diff
@@ -232,9 +239,14 @@ def generalized_advantage_estimation(rewards,
...
```
Comment (hnyu) [inline alf/utils/value_ops.py]: also remove the checks here?

File: alf/utils/value_ops.py (hunk 7 above)
Comment (hnyu) [reply alf/utils/value_ops.py]: It's still there?

Comment (emailweixu) [reply alf/utils/value_ops.py]: Sorry, only removed one place.

**Review by hnyu — APPROVED** [review]
